packages/fx-core/src/component/generator/utils.ts

```diff
@@ -230,8 +234,69 @@ export function getSampleRelativePath(sampleName: string): string {
   return `${sampleConfig.baseFolderName}/${sampleName}/`;
 }
 
+export function getSampleUrl(sample: SampleInfo): string {
+  if (sample.url === sampleConfig.baseUrl) {
+    return `${sample.url}${sample.id}`;
+  } else {
+    return sample.url;
+  }
+}
+
 export function zipFolder(folderPath: string): AdmZip {
   const zip = new AdmZip();
   zip.addLocalFolder(folderPath);
   return zip;
 }
+
+export async function downloadDirectory(
+  sampleUrl: string,
+  dstPath: string,
+  concurrencyLimits = sampleConcurrencyLimits,
+  retryLimits = sampleDefaultRetryLimits
+): Promise<void> {
+  //step 1: get file info
+  const urlParserRegex = /https:\/\/github.com[/]([^/]+)[/]([^/]+)[/]tree[/]([^/]+)[/](.*)/;
```

Comment:
Can be simplify to `const urlParserRegex = /https:\/\/github.com\/([^/]+)\/([^/]+)\/tree\/([^/]+)\/(.*)/;`